src/Stan/Analysis/Analyser.hs

```diff
@@ -133,6 +137,59 @@ analyseLazyFields insId hie =
     matchField :: HieAST TypeIndex -> Slist RealSrcSpan
     matchField = createMatch lazyField hie
 
+{- | Check for occurrences of pattern matching on @_@ for sum types (except
+literals).
+-}
+analysePatternMatch_ :: Id Inspection -> HieFile -> Observations
+analysePatternMatch_ insId hie =
+    mkObservation insId hie <$> analyseAstWith matchPatternMatch hie
+  where
+    matchPatternMatch :: HieAST TypeIndex -> Slist RealSrcSpan
+    matchPatternMatch node = memptyIfFalse
+        -- return empty list if it's not a case or lambda case
+        (hieMatchPatternAst hie node $ lambdaCase ||| case')
+        -- get list of all case branches
+        $ case nodeChildren node of
+              -- no branches = not observations
+              []     -> mempty
+              -- lambda case, first kid is pattern matching
+              [pm]   -> analyseBranches pm
+              -- case, first kid is @case exp of@, the second is pattern matching
+              _:pm:_ -> analyseBranches pm
+
+    {- Check the pattern matching child on some particular expressions.
+
+    -}
+    analyseBranches :: HieAST TypeIndex -> Slist RealSrcSpan
+    analyseBranches pm = case nodeChildren pm of
+        -- if there is no children = no observations
+        [] -> mempty
+        -- we need to check first and all other children separately
+        -- see 'firstChild' comment to understand the first child rules.
+        c:cs -> memptyIfFalse (firstPatternMatch c) $
+            {- if the first child satisfies rules of the first pattern matching
+            branch, then we need to find the child with pattern matching on @_@.
+            If there is no such expression = all is good.
+            -}
+            case find (\x -> hieMatchPatternAst hie x (patternMatch_ (?))) cs of
+                Nothing -> mempty
+                Just e  -> S.one (nodeSpan e)
+
+    {- The first pattern matching branch should not:
+    1. Be empty (makes no sense)
+    2. Be a literal pattern matching (e.g. on 'Int's or 'String's)
+    In all other cases we can continue our matching checks with other children.
+    -}
+    firstPatternMatch :: HieAST TypeIndex -> Bool
+    firstPatternMatch c = hieMatchPatternAst hie c patternMatchBranch &&
+        case takeWhile (\n -> hieMatchPatternAst hie n $ neg $ patternMatchArrow (?)) $ nodeChildren c of
```

Comment:
This `case ... of` is a bit long.. Maybe we can move it into separate variable? Like:

```haskell
let patterns = takeWhile ...
in case patterns of ...
```

Not sure, it will make the line shorter, but maybe will make code slightly more readable...